### PR TITLE
[WB-5045] 📼 Rewind 

### PR DESCRIPTION
**Pull request by dannygoldstein**
https://wandb.atlassian.net/browse/WB-5045

Description
-----------

Original problem statement: 

>When I restart an experiment, wandb will start dropping samples generated since the last experiment snapshot. What I'd want instead is to overwrite these samples with the new data generated by the experiment after restart (dropping the old data past the most recent snapshot). The reason this is important is because it's the new data that will impact the future, so not the old data. Is there a way to do that in wandb?

>Suppose my experiment was running for 5000 steps and then it crashed. However, I only saved it state when it finished 4000 steps. It is then restored at 4000 samples and it generates a bunch of metrics between steps 4000 and 5000 using wandb.log . However those datapoints are dropped, because they were already written by original experiment before restart. I'd rather have the old samples be replaced by new ones.

This PR implements the above. A related backend pr implements the associated backend changes 

Testing
-------

No tests yet, @raubitsj would appreciate your suggestions on best ways to test.

### CODE REVIEW

File: wandb/sdk/internal/internal_api.py [hunk 1]
```diff
@@ -887,49 +889,74 @@ def upsert_run(
...
+            try:
+                response = self.gql(mutation, variable_values=variable_values, **kwargs)
+            except CommError as e:
+                if "rewind" in e.args[0].lower():
+                    if rewind_step is not None:
+                        raise CommError(
+                            "Run rewinding is not supported on this version of the "
+                            "wandb backend. Please retry your request without rewinding."
+                        )
+                    continue
```
Comment (vanpelt) [inline wandb/sdk/internal/internal_api.py:945]: This is nasty enough I think it's time to wire through the `serverInfo` version information and do this based on the version of the server...

Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py:945]: That would be great!

**Comment by raubitsj** [comment]
Why add `rewind_step` to wandb.init()?

My proposed design was that the user could rely on wandb to determine what to rewind to, which is much harder to do when rewind_step is in the call to wandb.init()

For example:
```
run = wandb.init(resume=my_id)
# look up in your run when the last checkpoint was
last_checkpoint = run.summary.checkpoint_step
run.rewind(step=last_checkpoint)
```

File: wandb/sdk/internal/internal_api.py (hunk 1 above)
Comment (raubitsj) [reply wandb/sdk/internal/internal_api.py:945]: If someone wants to do that, it needs to be in the internal process (not the way it is currently implemented), and there are implications in doing that before the first upsert run call.  (new places for network hangs)

**Comment by dannygoldstein** [comment]
@raubitsj I added it to `init` so that it could share the upsert run call with `init/resume` and not have to have a separate recordloop Record type and path for handling rewind requests. The rewind info needs to get sent to the backend so that the backend can truncate the associated history. If you prefer I can add in that path, basically creating a new RunRecord in rewind and adding that to the queue

**Comment by dannygoldstein** [comment]
@raubitsj Re: testing, it would be great to test that the history truncation works as expected. This is going to be tough to mock out, I suppose we could add something to the mock server, but it would be better to test against a functioning backend. Is it possible to define a test in the new framework that only runs against a real live server?

**Comment by raubitsj** [comment]
> @raubitsj I added it to `init` so that it could share the upsert run call with `init/resume` and not have to have a separate recordloop path (and maybe Record type?) for handling rewind requests. The rewind info needs to get sent to the backend so that the backend can truncate the associated history. If you prefer I can add in that path, basically creating a new RunRecord in rewind and adding that to the queue

if you are using the wandb.init() call `rewind` is the wrong name now.  maybe it is resume_step.

File: wandb/sdk/internal/internal_api.py (hunk 1 above)
Comment (vanpelt) [reply wandb/sdk/internal/internal_api.py:945]: Given that's going to be a big chore, let's add a big comment saying this needs to be refactored when we pipe it through.

Comment (vanpelt) [reply wandb/sdk/internal/internal_api.py:945]: @raubitsj I had thought `wandb.login` (which is automatically called by wandb.init) was making a request already that could pipe the data through...

Comment (raubitsj) [reply wandb/sdk/internal/internal_api.py:945]: nope.  in `wandb.setup()` and it is happening in the user context with a 5 second timeout.
You can not rely on this.

**Comment by dannygoldstein** [comment]
closing in favor of #2794

**Pull request closed without merging**